### DISCUSSION TITLE
How to use json filter correctly?

### DISCUSSION THREAD

**Discussion by 1011382654** [directus/directus#17508] [Archive]
Expected results:
`
{
            "id": 1,
            "data": {
                "name": "text1",
                "value": 1
            }
 }
`

Equivalent to PostgreSQL : 
`SELECT * FROM demo WHERE data->>'name' = 'text1';`

Attempt based on the document failed to get the correct answer

https://github.com/directus/directus/pull/15889
https://docs.directus.io/reference/query.html

![image](https://user-images.githubusercontent.com/26688758/218647144-5c438477-6aba-4a70-8d05-05377751713a.png)
![image](https://user-images.githubusercontent.com/26688758/218647235-f7ed059d-55bb-423a-93e0-f789efc2eeb6.png)

**Comment by henk23** [comment]
Have not tried this, but based on your schema, this might work: `/items/demo?deep[data][_filter][data][name][_eq]=test1`

**Reply by 1011382654** [reply]
> Have not tried this, but based on your schema, this might work: `/items/demo?deep[data][_filter][data][name][_eq]=test1`
The query condition is not effective
![image](https://user-images.githubusercontent.com/26688758/218683774-fcd122d6-89a0-4999-9f6b-7a013d6d262d.png)

**Comment by br41nslug** [comment]
In your example that would be 
```
/items/demo?fields=*&filter={"json(data$.name)":{"_eq":"text2"}}
```

This feature has not been released yet tho

Note:
- `filter` is used to filter toplevel directus records based on a JSON value
- `deep` is used to filter the content inside a JSON array field

**Reply by 1011382654** [reply]
> In your example that would be
> 
> ```
> /items/demo?fields=*&filter={"json(data$.name)":{"_eq":"text2"}}
> ```
> 
> This feature has not been released yet tho
> 
> Note:
> 
> * `filter` is used to filter toplevel directus records based on a JSON value
> * `deep` is used to filter the content inside a JSON array field


Can it be understood that similar queries cannot be implemented?
`SELECT * FROM demo WHERE data->>'name' = 'text1';`

**Reply by br41nslug** [reply]
> Can it be understood that similar queries cannot be implemented?

How do you mean? Those 2 are equivalent.

**Reply by 1011382654** [reply]
> > 是不是可以理解为类似查询无法实现？
> 
> 你是什​​么意思？这2个是等价的。


![image](https://user-images.githubusercontent.com/26688758/218726385-b86c9333-fa3e-43cd-92cf-77bf232ca234.png)

Is the setting wrong?

![image](https://user-images.githubusercontent.com/26688758/218727830-6e854ad5-1662-4c91-92fd-73425e503ad2.png)
![image](https://user-images.githubusercontent.com/26688758/218727670-2c25779d-a716-4fe7-9eef-a5fd390819fa.png)

**Comment by br41nslug** [comment]
Is the database + version you're using in the compatibility chart for `"filter"`? https://github.com/directus/directus/pull/15889

Are you getting correct results for `/items/demo?fields=json(data$.name)`? Or does it work when adding an `alias` which should be the go-to anyhow.

**Reply by 1011382654** [reply]
> Is the database + version you're using in the compatibility chart for `"filter"`? #15889
> 
> Are you getting correct results for `/items/demo?fields=json(data$.name)`? Or does it work when adding an `alias` which should be the go-to anyhow.

My database is the recently installed docker postgis/postgis:latest

![image](https://user-images.githubusercontent.com/26688758/218730652-c7a0aec5-247e-4864-ad93-20ec0eb24b6c.png)

**Reply by br41nslug** [reply]
First of thanks found a bug in the "simplified" filter logic for the latest postgres 😄 

A couple things jump out that should not happen with a plain json structure:
- for one you shouldn't be getting 2 separate records as one array, json queries are not set up to do that
- it should have a generated field name instead of `data` because you're not using an alias
- it looks like you're getting the `value` instead of the `name`
🤔 

What i'd expect to see is this (just tested with postgres 15 locally for a schema like you describe):
![image](https://user-images.githubusercontent.com/9389634/218736026-b3887621-4c54-48bc-8c93-3e1888e5ea8f.png)

**Reply by 1011382654** [reply]
Looks like I getting the id

![image](https://user-images.githubusercontent.com/26688758/218828728-3b2fd966-7b54-458f-8edc-c499b3fe1336.png)

I'm sorry that I'm still in PostgreSQL, and I'm not familiar with how to set up the data model.
If possible, I want to store data and json queries like the following example

![image](https://user-images.githubusercontent.com/26688758/218832047-176a0242-8c3a-4aef-ab80-5f91ca591a4c.png)

**Reply by rijkvanzanten** [reply]
Are you using the open PR branch from a local build? The linked PR is open https://github.com/directus/directus/pull/15889 and not yet merged / released, so the `json()` filter isn't expected to work yet in the production build.

**Reply by 1011382654** [reply]
Yes, I am not running the development version.
I saw it on the branch. I think I know what to do.

![image](https://user-images.githubusercontent.com/26688758/218929155-bc6ad473-82ce-43fb-b78f-4490ca888a2d.png)

https://docs.directus.io/contributing/running-locally.html
